pkg/auth/authenticator/password/ldappassword/ldap.go

```diff
@@ -0,0 +1,240 @@
+func (a *Authenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	identity, ok, err := a.getIdentity(username, password)
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
+
+	user, err := a.mapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error creating or updating mapping for: %#v due to %v", identity, err)
+	}
+
+	return user, true, nil
+
+}
+
+// connect returns an established ldap connection, or an error if the connection could not be made (or successfully upgraded to TLS)
+// if no error is returned, the caller is responsible for closing the connection
+func (a *Authenticator) connect() (*ldap.Conn, error) {
+	switch a.options.URL.Scheme {
+	case ProtocolLDAP:
+		l, err := ldap.Dial("tcp", a.options.URL.Host)
+		if err != nil {
+			return nil, err
+		}
+
+		// If an insecure connection is desired, we're done
+		if a.options.Insecure {
+			return l, nil
+		}
+
+		// Attempt to upgrade to TLS
+		if err := l.StartTLS(a.options.TLSConfig); err != nil {
+			// We're returning an error on a successfully opened connection
+			// We are responsible for closing the open connection
+			l.Close()
+			return nil, err
+		}
+
+		return l, nil
+
+	case ProtocolLDAPS:
+		return ldap.DialTLS("tcp", a.options.URL.Host, a.options.TLSConfig)
+
+	default:
+		return nil, fmt.Errorf("unsupported scheme %q", a.options.URL.Scheme)
+	}
+}
+
+// getIdentity looks up a username in an LDAP server, and attempts to bind to the user's DN using the provided password
+func (a *Authenticator) getIdentity(username, password string) (authapi.UserIdentityInfo, bool, error) {
+	if len(username) == 0 || len(password) == 0 {
+		return nil, false, nil
+	}
+
+	// Make the connection
+	l, err := a.connect()
+	if err != nil {
+		return nil, false, err
+	}
+	defer l.Close()
+
+	// If specified, bind the username/password for search phase
+	if len(a.options.BindDN) > 0 {
+		if err := l.Bind(a.options.BindDN, a.options.BindPassword); err != nil {
+			return nil, false, err
+		}
+	}
+
+	// & together the filter specified in the LDAP options with the user-specific filter
+	filter := fmt.Sprintf("(&%s(%s=%s))",
+		a.options.URL.Filter,
+		ldap.EscapeFilter(a.options.URL.QueryAttribute),
+		ldap.EscapeFilter(username),
+	)
+
+	// Build list of attributes to retrieve
+	attrs := util.NewStringSet(a.options.URL.QueryAttribute)
+	attrs.Insert(a.options.AttributeEmail...)
+	attrs.Insert(a.options.AttributeName...)
+	attrs.Insert(a.options.AttributePreferredUsername...)
+	attrs.Insert(a.options.AttributeID...)
+
+	// Search for LDAP record
+	searchRequest := ldap.NewSearchRequest(
+		a.options.URL.BaseDN,     // base dn
+		int(a.options.URL.Scope), // scope
+		ldap.NeverDerefAliases,   // deref
+		2,            // size limit, we want to know if this is not unique, but don't want the entire tree
+		0,            // time limit
```

Comment:
`0` is unlimitted?  Seems like we'd want to restrict that to something reasonable.